src/ints/mouse_vmware.cpp

```diff
@@ -0,0 +1,208 @@
+    default:
+        LOG_WARNING("Mouse: unimplemented VMware subcommand 0x%08x", reg_ebx);
+        break;
+    }
+}
+
+static uint16_t PortRead_VMW(io_port_t, io_width_t) {
+    if (reg_eax != VMW_MAGIC)
+        return 0;
+
+    switch (reg_cx) {
+    case VMW_CMD::GETVERSION:         CmdGetVersion();        break;
+    case VMW_CMD::ABSPOINTER_DATA:    CmdAbsPointerData();    break;
+    case VMW_CMD::ABSPOINTER_STATUS:  CmdAbsPointerStatus();  break;
+    case VMW_CMD::ABSPOINTER_COMMAND: CmdAbsPointerCommand(); break;
+    default:
+        LOG_WARNING("Mouse: unimplemented VMware command 0x%08x", reg_ecx);
+        break;
+    }
+
+    return reg_ax;
+}
+
+bool MouseVMW_NotifyMoved(int32_t x_abs, int32_t y_abs) {
+    float vmw_x, vmw_y;
+    if (mouse_video.fullscreen) {
+        // We have to maintain the diffs (offsets) between host and guest
+        // mouse positions; otherwise in case of clipped picture (like
+        // 4:3 screen displayed on 16:9 fullscreen mode) we could have
+        // an effect of 'sticky' borders if the user moves mouse outside
+        // of the guest display area
+
+        if (x_abs + offset_x < mouse_video.clip_x)
+                offset_x = static_cast<int16_t>(mouse_video.clip_x - x_abs);
+        else if (x_abs + offset_x >= mouse_video.res_x + mouse_video.clip_x)
+                offset_x = static_cast<int16_t>(mouse_video.res_x + mouse_video.clip_x - x_abs - 1);
+
+        if (y_abs + offset_y < mouse_video.clip_y)
+                offset_y = static_cast<int16_t>(mouse_video.clip_y - y_abs);
+        else if (y_abs + offset_y >= mouse_video.res_y + mouse_video.clip_y)
+                offset_y = static_cast<int16_t>(mouse_video.res_y + mouse_video.clip_y - y_abs - 1);
+
+        vmw_x = static_cast<float>(x_abs + offset_x - mouse_video.clip_x);
+        vmw_y = static_cast<float>(y_abs + offset_y - mouse_video.clip_y);
+    }
+    else {
+        vmw_x = static_cast<float>(std::max(x_abs - mouse_video.clip_x, 0));
+        vmw_y = static_cast<float>(std::max(y_abs - mouse_video.clip_y, 0));
+    }
+
+    auto old_x = scaled_x;
+    auto old_y = scaled_y;
+
+    scaled_x = static_cast<uint16_t>(std::min(0xffffu,
+        static_cast<uint32_t>(vmw_x * 0xffff / static_cast<float>(mouse_video.res_x - 1) + 0.499)));
+    scaled_y = static_cast<uint16_t>(std::min(0xffffu,
+        static_cast<uint32_t>(vmw_y * 0xffff / static_cast<float>(mouse_video.res_y - 1) + 0.499)));
+
+    updated = true;
+
+    return mouse_vmware && (old_x != scaled_x || old_y != scaled_y);
+}
+
+void MouseVMW_NotifyPressedReleased(uint8_t buttons_12S) {
+    buttons_vmw = 0;
+
+    if (buttons_12S & 1) buttons_vmw |=VMW_BUTTON::LEFT;
+    if (buttons_12S & 2) buttons_vmw |=VMW_BUTTON::RIGHT;
+    if (buttons_12S & 4) buttons_vmw |=VMW_BUTTON::MIDDLE;
+
+    updated = true;
+}
+
+void MouseVMW_NotifyWheel(int32_t w_rel) {
+    if (mouse_vmware) {
+        wheel   = static_cast<int8_t>(std::clamp(w_rel + wheel, -0x80, 0x7f));
```

Comment:
After all these comments, you should have your clamping functions implemented as helper functions at this point 😀